## Домашнее задание к 4 уроку (математика, теория вероятностей)

Easy
Вычислить вероятность выживания, если пассажиру:

от 0 до 25 лет
от 25 до 50 лет
от 50 лет и старше

Medium

Вычислить с какой вероятностью пассажир женского пола, если известно, что он выжил (использовать теорему Байеса)

Hard

Закон распределения задается функцией плотности

𝑓(𝑥)=𝜆𝑒−𝜆𝑥,𝑥>0 
𝑓(𝑥)=0,𝑥<=0 
Найти вероятность того, что событие произойдет при  𝑥∈[1,2] 
P.S. в scipy есть функция для интегрирования функции

## Ответы:

In [1]:
import pandas as pd  # Импортируем pandas

# 1. Задача Easy
Вычислить вероятность выживания, если пассажиру:

от 0 до 25 лет
от 25 до 50 лет
от 50 лет и старше

In [2]:
titanic = pd.read_csv('./data/train.csv')
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Для определения вероятности выживания, сначала считаем общее число пассажиров по PassengerID, 
присваиваем значение переменной people_total

In [3]:
people_total = titanic.PassengerId.count()    
print('Число пассажиров всего, человек = ', people_total)

Число пассажиров всего, человек =  891


Cчитаем число пассажиров меньше 25 лет, от 25 до 50, старше 50
Записываем в переменные.

In [4]:
people_younger_25 = titanic[titanic.Age < 25].PassengerId.count()

In [5]:
people_older_25_50 = titanic[(titanic.Age >= 25) & (titanic.Age < 50)].PassengerId.count()                          

In [6]:
people_older_50 = titanic[titanic.Age >= 50].PassengerId.count()

Создаем новый датафрейм только с выжившими

In [7]:
surviveddf = titanic[titanic.Survived == 1] 

Считаем выживших по группам до 25 лет, 25-50 лет, старше 50

In [8]:
people_younger_25_alive = surviveddf[surviveddf.Age < 25].Survived.count()

In [9]:
people_older_25_50_alive = surviveddf[(surviveddf.Age >= 25) & (surviveddf.Age < 50)].Survived.count()

In [10]:
people_older_50_alive = surviveddf[surviveddf.Age >= 50].Survived.count()

Для наглядности формируем датафрейм (таблицу) со всеми пассажирами и с выжившими пассажирами в разбивке по возрастам, 
с учетом определенных выше переменных.
В столбце "Survival probability, %" в %ах отражаем вероятность выжить по каждой группе возрастов
(отношение выживших по возрастным группам к полной численности каждой возрастной группы).

In [11]:
alive_df = pd.DataFrame(['younger 25', '25-50', 'older 50'])

In [12]:
alive_df['All passengers divided by Ages'] = pd.Series([people_younger_25, people_older_25_50, people_older_50])

In [13]:
alive_df['Survived passengers divided by Ages'] = pd.Series([people_younger_25_alive, people_older_25_50_alive, people_older_50_alive])

In [14]:
alive_df['Survival probability, %'] = (alive_df['Survived passengers divided by Ages']/alive_df['All passengers divided by Ages'])*100

In [15]:
alive_df  # выведем сформированный датафрейм с вероятностью выживания

0  All passengers divided by Ages  \
0  younger 25                             278   
1       25-50                             362   
2    older 50                              74   

   Survived passengers divided by Ages  Survival probability, %  
0                                  118                42.446043  
1                                  145                40.055249  
2                                   27                36.486486

Самая высокая вероятность выжить в разбивке по группам возрастов - в группе от 0 до 25 лет = 42,446%. В группе 25-50 лет она незначительно снижается - до 40,055%, и для лиц старше 50 составляет уже 36,486%. Но в целом вероятность выживания в разбивке по возрастным группам (до 25, 25-50, старше 50) не сильно различается.

# 2. Задача Medium

Вычислить с какой вероятностью пассажир женского пола, если известно, что он выжил (использовать теорему Байеса)

Ответ:
у нас все пассажиры на борту мужского и женского пола. Для начала считаем доли пассажиров женского и мужского пола на борту.
Записываем расчетные показатели в переменные, которые потребуются для дальнейших расчетов (см. комментарии ниже)

In [16]:
female_passengers = titanic[titanic.Sex == 'female'].Sex.count()
print('Число женщин на борту: ', female_passengers)

Число женщин на борту:  314


In [17]:
male_passengers = titanic[titanic.Sex == 'male'].Sex.count()
print('Число мужчин на борту: ', male_passengers)

Число мужчин на борту:  577


In [18]:
part_of_female = female_passengers / people_total 
part_of_male = male_passengers / people_total 
print('Доля женщин на борту: ', part_of_female)
print('Доля мужчин на борту: ', part_of_male)

Доля женщин на борту:  0.35241301907968575
Доля мужчин на борту:  0.6475869809203143


Решение:
Первая часть решения состоит в использовании формулы полной вероятности. Иными словами, вычисления проводятся в предположении, 
что испытание ещё не произведено и событие «пассажир выжил" пока не наступило. Возможны две гипотезы:
1) пассажир будет из 1-й группы - женщин;
2) пассажир будет из 2-й группы - мужчин;
 Всего: 314 + 577 = 891 пассажиров на борту. По классическому определению:
P_B1 = 314 / 891 = 0.35   P_B2 = 577 / 891 = 0.65  
Контроль:  0.35 + 0.65 = 1

In [19]:
P_B1 = part_of_female
P_B2 = part_of_male
print('P_B1= ', part_of_female)
print('P_B2= ', part_of_male)

P_B1=  0.35241301907968575
P_B2=  0.6475869809203143


Для удобства расчетов и определения долей выживших женщин и мужчин в общем числе каждой группы - мужчин и женщин -
формируем датафрейм с выжившими и на базе его считаем выживших и их доли

In [20]:
female_passengers_alive = surviveddf[surviveddf.Sex == 'female'].Survived.count()

In [21]:
male_passengers_alive = surviveddf[surviveddf.Sex == 'male'].Survived.count()

In [22]:
female_part_alive = (female_passengers_alive / female_passengers)
male_part_alive = (male_passengers_alive / male_passengers)
print('Число выживших женщин: ', female_passengers_alive)
print('Число выживших мужчин: ', male_passengers_alive)
print('Доля выживших женщин в общем количестве женщин: ', female_part_alive)
print('Доля выживших мужчин в общем количестве мужчин: ', male_part_alive)

Число выживших женщин:  233
Число выживших мужчин:  109
Доля выживших женщин в общем количестве женщин:  0.7420382165605095
Доля выживших мужчин в общем количестве мужчин:  0.18890814558058924


Для удобства можно отобразить используемые данные в таблице:

In [23]:
survived_male_female = pd.DataFrame(['female', 'male'])
survived_male_female['total (survived and dead)'] = pd.Series([female_passengers, male_passengers])
survived_male_female['part in all passengers'] = pd.Series([part_of_female, part_of_male])
survived_male_female['survived'] = pd.Series([female_passengers_alive, male_passengers_alive])
survived_male_female['part of survived in group'] = pd.Series([female_part_alive, male_part_alive])
survived_male_female

0  total (survived and dead)  part in all passengers  survived  \
0  female                        314                0.352413       233   
1    male                        577                0.647587       109   

   part of survived in group  
0                   0.742038  
1                   0.188908

Рассмотрим зависимое событие:  –  пассажир выжил.
В первой группе - женщин - 0,7420 выживших, поэтому P_B1_A = 233/314 = 0,7420 вероятность того, 
что пассажир выжил при условии, что он женщина.

Аналогично, во второй группе - мужчин - 0,1889 выживших и  – вероятность того, что пассажир выжил при условии при условии,
что он мужчина, P_B2_A = 109 / 577 = 0,1889.

По формуле полной вероятности:
P_A = P_B1 * P_B1_A + P_B2 * P_B2_A
– вероятность того, что пассажир выжил.

In [24]:
P_B1_A = female_part_alive
P_B2_A = male_part_alive
print('P_B1_A = ', female_part_alive)
print('P_B2_A = ', male_part_alive)

P_B1_A =  0.7420382165605095
P_B2_A =  0.18890814558058924


In [25]:
P_A = P_B1*P_B1_A + P_B2*P_B2_A

In [26]:
print("Вероятность того, что пассажир выжил =", P_A)

Вероятность того, что пассажир выжил = 0.3838383838383838


Часть вторая. Пусть пассажир выжил, то есть событие A произошло. 
То есть мы констатируем тот факт, что событие  произошло. По формулам Байеса:

PA_B1 = (P_B1 * P_B1_A) / P_A
- вероятность того, что выживший пассажир женщина

PA_B2 = (P_B2 * P_B2_A) / P_A
- вероятность того, что выживший пассажир мужчина

Проверка: PA_B1 + PA_B2 = 1

In [27]:
PA_B1 = (P_B1 * P_B1_A) / P_A
print("Вероятность того, что выживший пассажир ЖЕНЩИНА =", PA_B1)

Вероятность того, что выживший пассажир ЖЕНЩИНА = 0.6812865497076023


In [28]:
PA_B2 = (P_B2 * P_B2_A) / P_A
print("Вероятность того, что выживший пассажир мужчина =", PA_B2)

Вероятность того, что выживший пассажир мужчина = 0.31871345029239767


На всякий случай проверим корректность полученных результатов:

In [29]:
if round(PA_B1 + PA_B2, 2) == 1:
    print('корректно')
else:
    print('некорректно')

корректно


In [30]:
print("ОТВЕТ: Вероятность того, что выживший пассажир ЖЕНЩИНА составляет, c учетом округлений:", 
      round(PA_B1, 4), 'или', round(PA_B1*100, 2), '%')

ОТВЕТ: Вероятность того, что выживший пассажир ЖЕНЩИНА составляет, c учетом округлений: 0.6813 или 68.13 %
